### PR TITLE
COLDBOX-749 singleton renderer

### PR DESCRIPTION
**Pull request by DominicWatson**
The renderer service is current a transient service. This makes it threadsafe but also make it slow (if you have a request with lots of small renderView() calls, the request time adds up).

The following changes allow the renderer service to be a singleton while still being threadsafe for variables set within included views.

It does this by creating a shallow clone of itself just for the method that includes the cfm view template. This performs well and does not modify the scope of the singleton instance. Shallow cloning is fast.

### CODE REVIEW

**Comment by bdw429s** [comment]
@DominicWatson Did you see my question on the JIRA ticket?

**Comment by bdw429s** [comment]
Also, shallow duplicate is a Lucee feature as you can see by the failing builds :)

**Comment by DominicWatson** [comment]
Ha ha, yep - just coming on here to comment to that effect!

**Comment by bdw429s** [comment]
I think `StructCopy()` will give you what you want on both engines.

**Comment by DominicWatson** [comment]
Cool, was going to try that. I'll run it and see what happens. FWIW, We've been running this modification for over 4 years in Lucee without any probs (in scores of different applications).

**Comment by DominicWatson** [comment]
n.b. StructCopy() works in Lucee, but not in ACF - will need to look into how this can work on ColdFusion before it could be merged.

**Comment by DominicWatson** [comment]
So, there is no shallow copy of a CFC possible in ACF. Latest commit checks if Lucee and does StructCopy() if so. If not lucee, the code does an inline `new Renderer( controller, htmlHelper )` and sets other normally wirebox injectable variables.

My rough load testing (looping over 10,000 renderView() calls) shows:

Lucee: a little more than twice as fast
ACF: around 150% faster

**Comment by lmajano** [comment]
Did you see my comments about the explicit view data? This will become unsafe.

**Comment by GunnarLieb** [comment]
@lmajano are the comments public? I find the proposed changes very interesting and I would like to follow the discussion.

**Comment by lmajano** [comment]
I had forgot to submit the review

File: system/web/Renderer.cfc [hunk 1]
```diff
@@ -332,61 +343,47 @@ component accessors="true" serializable="false" extends="coldbox.system.Framewor
...
 			if( arguments.collectionMaxRows NEQ 0 AND arguments.collectionMaxRows LTE recLen ){ recLen = arguments.collectionMaxRows; }
 			// Create local marker
-			variables._items	= recLen;
+			viewArgs._items	= recLen;
 			// iterate and present
 			for( x=arguments.collectionStartRow; x lte recLen; x++ ){
 				// setup local variables
-				variables._counter  = x;
-				variables[ arguments.collectionAs ] = arguments.collection[ x ];
+				viewArgs._counter  = x;
```
[the remark this answers is not in the document]
Comment (GunnarLieb) [reply system/web/Renderer.cfc]: Is there a test case or an example where it breaks?

**Comment by DominicWatson** [comment]
I think it would make sense to refactor to some kind of rendererFactory so that there is less refactoring in the Renderer.cfc itself. When you do `renderView()`, the helper would do `RendererFactory.getRenderer().renderView(...)`.

The `viewArgs` piece was necessary to not do request specific logic in the variables scope - but I'll refactor.

**Comment by DominicWatson** [comment]
@lmajano have refactored with backward compatibility in mind. You are now able to still call getInstance( "renderer@coldbox" ) and no internal variable name references have changed.

Instead, there is a separate factory object that deals with getting new renderer instances in a faster way.

**Comment by lmajano** [comment]
Hi @DominicWatson, taking time on this as as it stands, it breaks ContentBox as ContentBox expects the renderer to be created by WireBox.  So anything expecting the Renderer to be created by WireBox will fail because they listen to the afterInstanceCreation

**Comment by lmajano** [comment]
Sorry, @DominicWatson this has taken me very long to get to. Still working on it.

**Comment by lmajano** [comment]
I haven't forgotten about you @DominicWatson

**Comment by lmajano** [comment]
Thanks for all the hard work on this @DominicWatson we reused lots of your commit on this!  Thanks always for your support and time!

**Comment by DominicWatson** [comment]
:+1: glad to start the discussion. Love the quirkishness of the final approach.

**Pull request closed without merging**